### ISSUE TITLE
Is there a better option for pretty printing json?

### ISSUE THREAD

**Issue by nodesocket** [nodesocket/jsonlite#3]
Piping to `python -m json.tool` to pretty print json is super expensive. Is there a better option? Is a native bash implementation even possible? 

See: https://github.com/nodesocket/jsonlite/blob/master/jsonlite.sh#L34

**Comment by skibz** [comment]
@nodesocket have you seen/considered [jq](https://github.com/stedolan/jq) ?
it outputs pretty printed json by default and even supports transformations on the data it's given based on input expressions.

really love the look of this little tool
keep on truckin' :+1: :tada:

**Comment by hl** [comment]
It's a better idea to put this as an option (flag) because it would only be useful during dev. In that case it really doesn't matter how expensive it is.

**Comment by kylehill** [comment]
Since it's JS-based, might node be cheaper?

node -e "process.stdout.write(JSON.stringify(" + obj + ", null, 2))"

**Comment by skibz** [comment]
@kylehill 
pretty sure you'd have to interpolate the bash variable into the `node -e` call, like:

``` bash
node -e "process.stdout.write(JSON.stringify(${2}, null, 2));"
```

but, in any case, the system is bash (not node) based.

**Comment by nodesocket** [comment]
@skibz jq looks great, but I am concerned about the dependency addition.

@zenry pretty print is actually really nice and useful all the time. When you call get, the document comes back pretty printed which is good UX.

**Comment by nodesocket** [comment]
@kylehill and @skibz in terms of using node, I think it is still going to be slow, probably even slower than python because of the overhead of starting up `v8`.

**Comment by hl** [comment]
Not saying it's not nice, I'm saying that **if** something like this will be used as a storage backend then 'pretty print' doesn't really matter :smile: 

~~Maybe something like https://stedolan.github.io/jq/~~

**Comment by evuez** [comment]
YAJL's [json_reformat](https://github.com/lloyd/yajl/blob/master/reformatter/json_reformat.c) seems to be a great option, at least regarding performances (well in fact I don't think you can actually get any better regarding performances). On Fedora you get the `json_reformat` as a standalone, and there isn't any dependencies.

I put it in my [pull request](https://github.com/nodesocket/jsonlite/pull/7).

**Comment by nodesocket** [comment]
I think doing:

```
echo "$2" | jq '.' > "$CWD/jsonlite.data/$UUID"
```

will work. Anybody have any time to benchmark the difference between the above and:

```
echo "$2" | python -m json.tool > "$CWD/jsonlite.data/$UUID"
```

A simple test that calls `set` 1,000 times and timing it should work.

**Comment by josegonzalez** [comment]
I "benchmarked" this on my laptop using 5k iterations. Caveat is that I used `jsonlite` as a library function as opposed to shelling out, but that probably doesn't matter too much. The winner (not by much) is `json_reformat`. We can probably do an if block as I do in #9.

### json_reformat

```
real    0m52.686s
user    0m28.195s
sys     0m33.924s
```

### jq

```
real    0m58.094s
user    0m31.460s
sys     0m36.082s
```

### json.tool

```
real    2m48.762s
user    1m40.461s
sys     1m12.188s
```

**Comment by nodesocket** [comment]
Refactored and released version `0.6.0`. Now supports `json_reformat` and `jq`.

[https://github.com/nodesocket/jsonlite/releases/tag/0.6.0](https://github.com/nodesocket/jsonlite/releases/tag/0.6.0)
